src/codegen/x86_64/arch_x86_64.c

```diff
@@ -380,8 +383,1257 @@ static enum IndirectJumpType negate_jump(enum IndirectJumpType j) {
+
+static uint8_t modrm_byte(uint8_t mod, uint8_t reg, uint8_t rm) {
+  // Ensure no bits above the amount expected are set.
+  ASSERT((mod & (~0b11)) == 0);
+  // Top bit of register stored in REX bit(s), but may still be present here.
+  ASSERT((reg & (~0b1111)) == 0);
+  ASSERT((rm & (~0b1111)) == 0);
+  return (uint8_t)((mod << 6) | ((reg & 0b111) << 3) | rm);
+}
+
+static void mcode_imm_to_reg(CodegenContext *context, enum Instruction inst, int64_t immediate, RegisterDescriptor destination_register, enum RegSize size) {
+  if ((inst == I_SUB || inst == I_ADD) && immediate == 0) return;
+
+  switch (inst) {
+  case I_MOV: {
+
+    switch (size) {
+    case r8: {
+      // Move imm8 to r8
+      // 0xb0+ rb ib
+      uint8_t op = 0xb0 + rb_encoding(destination_register);
+      fwrite(&op, 1, 1, context->machine_code);
+      fwrite(&immediate, 1, 1, context->machine_code);
+    } break;
+    case r16: {
+      // Move imm16 to r16
+      // 0xb8+ rw iw
+      uint8_t op = 0xb8 + rw_encoding(destination_register);
+      fwrite(&op, 1, 1, context->machine_code);
+      fwrite(&immediate, 1, 2, context->machine_code);
+    } break;
+    case r32: {
+      // Move imm32 to r32
+      // 0xb8+ rd id
+      uint8_t op = 0xb8 + rd_encoding(destination_register);
+      fwrite(&op, 1, 1, context->machine_code);
+      fwrite(&immediate, 1, 4, context->machine_code);
+    } break;
+    case r64: {
+      // Move imm64 to r64
+      // REX.W + B8+ rd io
+      uint8_t rex = rexw_byte();
+      uint8_t op = 0xb8 + rd_encoding(destination_register);
+      fwrite(&rex, 1, 1, context->machine_code);
+      fwrite(&op, 1, 1, context->machine_code);
+      fwrite(&immediate, 1, 8, context->machine_code);
+    } break;
+
+    } // switch (size)
+
+  } break; // case I_MOV
+
+  case I_ADD:
+  case I_SUB: {
+
+    // Immediate add/sub both share the same opcodes, just with a different opcode extension in ModRM:reg.
+    uint8_t add_extension = 0;
+    uint8_t sub_extension = 5;
+    uint8_t modrm = 0;
+    uint8_t destination_regbits = regbits(destination_register);
+    // Mod == 0b11  ->  register
+    // Reg == Opcode Extension (5 for sub, 0 for add)
+    // R/M == Destination
+    if (inst == I_ADD)
+      modrm = modrm_byte(0b11, add_extension, destination_regbits);
+    else modrm = modrm_byte(0b11, sub_extension, destination_regbits);
+
+    switch (size) {
+    case r8: {
+      // 0x80 /5 ib
+      uint8_t op = 0x80;
+
+      // Encode a REX prefix if the ModRM register descriptor needs
+      // the bit extension.
+      if (REGBITS_TOP(destination_regbits)) {
+        uint8_t rex = rex_byte(false, false, false, REGBITS_TOP(destination_regbits));
+        fwrite(&rex, 1, 1, context->machine_code);
+      }
+
+      int8_t imm8 = (int8_t)immediate;
+
+      fwrite(&op, 1, 1, context->machine_code);
+      fwrite(&modrm, 1, 1, context->machine_code);
+      fwrite(&imm8, 1, 1, context->machine_code);
+    } break;
+    case r16: {
+      // 0x81 /5 iw
```

Comment:
Missing `0x66 + `

Reply:
Fixed as of [1234801](https://github.com/LensPlaysGames/Intercept/pull/66/commits/1234801a026b424dbe55d3b1541a35e70c6c88f0)